### PR TITLE
Implementing Graceful Shutdown for VMs

### LINKED ISSUE #73: Implement Graceful Shutdown for VMs
# Summary

Currently, libvirt-provider destroys VMs directly, potentially causing data loss or corruption as applications running inside the VM may not have adequate time for clean-up and a graceful shutdown.

It would be highly beneficial if libvirt-provider could support a graceful shutdown process for VMs. This process should involve sending an ACPI power event to the guest operating system within the VM. If the guest OS supports ACPI events, applications and processes inside the VM should close gracefully.

## Basic example

libvirt-provider could leverage libvirt's `virsh shutdown` feature to initiate the graceful shutdown and follow up with a `virsh destroy` command after a timeout period if the VM hasn't shut down.

Need to confirm if Gardenlinux supports ACPI and has the necessary ACPI handlers/daemons to process such a shutdown request.

## Motivation

Enhance data safety and integrity for applications running in VMs.

### PR DESCRIPTION
**Pull request by so-sahu**
Fixes https://github.com/ironcore-dev/libvirt-provider/issues/73

### CODE REVIEW

File: pkg/controllers/machine_controller.go
Comment (lukas016) [inline pkg/controllers/machine_controller.go:313]: will it requeue again?

Comment (so-sahu) [reply pkg/controllers/machine_controller.go:313]: yes, the control will go to line 198, and item will be requeued.
(https://github.com/ironcore-dev/libvirt-provider/blob/8c953a3bdf012e2ff2a4f6732027a21679b4e7cd/pkg/controllers/machine_controller.go#L198C3-L198C31)

File: pkg/controllers/machine_controller.go
Comment (lukas016) [inline pkg/controllers/machine_controller.go:384]:
```suggestion
	// Performance issue will be in edge cases, if machine ignore shutdown command (not booted,  ignoring acpi signals, ...).
```

Comment (so-sahu) [reply pkg/controllers/machine_controller.go:384]: ok. I will change it.

File: pkg/meta/meta.go
Comment (lukas016) [inline pkg/meta/meta.go:14]: it looks as mixing of 2 key format without _ and with _. I prefer unified naming convetion

**Comment by lukas016** [comment]
@lukasfrank if i remember correctly, you implemented exponential requeue for every returned error. Is it correct? Because if it is true, we can simplify code and delete requeue error and custom requeue interval.

**Comment by lukasfrank** [comment]
> @lukasfrank if i remember correctly, you implemented exponential requeue for every returned error. Is it correct? Because if it is true, we can simplify code and delete requeue error and custom requeue interval.

Correct - the requeuing is already implemented in the `RateLimitingQueue` ( `k8s.io/utils/clock`)

**Comment by so-sahu** [comment]
I acknowledge that the requeuing functionality is already incorporated in the RateLimitingQueue. However, I find the nanosecond precision too rapid, especially in the context of virtual machine shutdowns (refer to: [source](https://github.com/kubernetes/client-go/blob/fb8b7346aacefea5ee2ab2e234afc4451c90c435/util/workqueue/default_rate_limiters.go#L97)). When utilizing the RateLimitingQueue for VM shutdowns, the heightened precision leads to accelerated reconciliations, resulting in an unnecessary surge in the number of machine reconciliations.

Therefore, I propose the inclusion of existing custom requeue interval for the graceful shutdown of VMs.

**Comment by lukasfrank** [comment]
> I acknowledge that the requeuing functionality is already incorporated in the RateLimitingQueue. However, I find the nanosecond precision too rapid, especially in the context of virtual machine shutdowns (refer to: [source](https://github.com/kubernetes/client-go/blob/fb8b7346aacefea5ee2ab2e234afc4451c90c435/util/workqueue/default_rate_limiters.go#L97)). When utilizing the RateLimitingQueue for VM shutdowns, the heightened precision leads to accelerated reconciliations, resulting in an unnecessary surge in the number of machine reconciliations.
> 
> Therefore, I propose the inclusion of existing custom requeue interval for the graceful shutdown of VMs.

@so-sahu I think you mixed something up: The delay is only converted to nanoseconds. We use the following (default) settings: 

```
func DefaultControllerRateLimiter() RateLimiter {
	return NewMaxOfRateLimiter(
		NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second),
		// 10 qps, 100 bucket size.  This is only for retry speed and its only the overall factor (not per item)
		&BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
	)
}
```

File: pkg/controllers/machine_controller.go
Comment (lukas016) [inline pkg/controllers/machine_controller.go:45]: Do we need this reconciletime?

Comment (lukas016) [reply pkg/controllers/machine_controller.go:45]:
```text
5
10
20
40
80
160
320
640
1280 (1.28s)
2560
5120
10240
20480
40960
81920
163840
327680
655360
```
@lukasfrank as you can see, we currently start with 5ms timeout between reconcile loops. So if we will return error for reconcilation, it will call another reconcile loops under 1seconds what is quick for shutdown process because machine need minimal seconds for graceful shutdown. So sourav has problem with this first 10 reconciliation  loops.

**Comment by lukas016** [comment]
> > I acknowledge that the requeuing functionality is already incorporated in the RateLimitingQueue. However, I find the nanosecond precision too rapid, especially in the context of virtual machine shutdowns (refer to: [source](https://github.com/kubernetes/client-go/blob/fb8b7346aacefea5ee2ab2e234afc4451c90c435/util/workqueue/default_rate_limiters.go#L97)). When utilizing the RateLimitingQueue for VM shutdowns, the heightened precision leads to accelerated reconciliations, resulting in an unnecessary surge in the number of machine reconciliations.
> > Therefore, I propose the inclusion of existing custom requeue interval for the graceful shutdown of VMs.
> 
> @so-sahu I think you mixed something up: The delay is only converted to nanoseconds. We use the following (default) settings:
> 
> ```
> func DefaultControllerRateLimiter() RateLimiter {
> 	return NewMaxOfRateLimiter(
> 		NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second),
> 		// 10 qps, 100 bucket size.  This is only for retry speed and its only the overall factor (not per item)
> 		&BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
> 	)
> }
> ```

Hi, @lukasfrank and @so-sahu i did mistake when i asked this as comment. i created conversation for continue discussion about this problem (https://github.com/ironcore-dev/libvirt-provider/pull/101#discussion_r1448436933) because using standard comments impact readability of PR bad way.

**Comment by lukasfrank** [comment]
As part of an offline discussion (@so-sahu  @lukas016)  we can conclude that the current approach is misusing the reconcile loop to delay the deletion process which leads to odd and unstable 

The agreed second iteration behaves the following way: 
- [x]  Server adds the deletion timestamp (already there) 
- [ ]  Additional routine checks periodically ( `store.List`) if machines in the store contain deletion timestamp 
  - [ ] If timestamp + shutdown period is exceed the machine is being shutdown forcefully 
  - [ ] Otherwise VM receives graceful shutdown signal 
- [ ] machine dependent resources are being deleted + remove finalizer

To summarise: The shutdown handling is being factored out of the main routine + the main routine stays almost the same

**Review by lukasfrank — CHANGES_REQUESTED** [review]
As part of an offline discussion (@so-sahu  @lukas016)  we can conclude that the current approach is misusing the reconcile loop to delay the deletion process which leads to odd and unstable behaviour. 

The agreed second iteration behaves the following way: 
- [x]  Server adds the deletion timestamp (already there) 
- [ ]  Additional routine checks periodically ( `store.List`) if machines in the store contain deletion timestamp 
  - [ ] If timestamp + shutdown period is exceed the machine is being shutdown forcefully 
  - [ ] Otherwise VM receives graceful shutdown signal 
- [ ] machine dependent resources are being deleted + remove finalizer

To summarise: The shutdown handling is being factored out of the main routine + the main routine stays almost the same

File: provider/cmd/app/app.go
Comment (afritzler) [inline provider/cmd/app/app.go:122]:
```suggestion
	fs.DurationVar(&o.VMGracefulShutdownTimeout, "vm-graceful-shutdown-timeout", 5*time.Minute, "Duration to wait for the VM to gracefully shut down. If the VM does not shut down within this period, it will be forcibly terminated.")
```

**Review by afritzler — CHANGES_REQUESTED** [review]
Instead of requeuing the `Machine` when we shutdown the domain I would rather implement the following termination flow:
- When we delete a `Machine`, we add the `deletionTimestamp` to the `Machine`s metadata and initiate a domain shutdown
- In parallel we start in the `app.go` a GC go routine (e.g. a Garbage collection process) which configurable via flag (e.g. `--gc-resync-period`) is running over all `Machines` in the local store and checks if `deletionTimestamt + GracePeriod` > `currentTime`. If that is the case we forcefully shutdown the domain.

I would not reuse the requeuing mechanism in the reconciler for doing cleanup jobs.

UPDATE: sorry I didn't see @lukasfrank comment. But essentially this goes into the same direction.

File: pkg/controllers/machine_controller.go
Comment (afritzler) [inline pkg/controllers/machine_controller.go:427]: We should avoid logging which printing "successful" in the message. Instead we should use e.g. "removing XYZ" when initiating an action and "removed XYZ" when the process is finished. The definition of what "successful" means is most of the time ambiguous.

**Review by afritzler — CHANGES_REQUESTED** [review]

File: provider/cmd/app/app.go
Comment (lukas016) [inline provider/cmd/app/app.go:123]: what do you think about gc-vm-graceful-shutdown-period/timeout. Mainly i want to know gc-resync-interval and this argument have some relationship.

File: pkg/controllers/machine_controller.go
Comment (lukas016) [inline pkg/controllers/machine_controller.go:194]: i think we need recovery logic for panic. this function has to be design as failure tolerant

File: pkg/controllers/machine_controller.go
Comment (lukas016) [inline pkg/controllers/machine_controller.go:229]: rename function. cleanup isn't anymore part of this function

**Review by lukas016 — CHANGES_REQUESTED** [review]

File: pkg/controllers/machine_controller.go
Comment (lukas016) [reply pkg/controllers/machine_controller.go:194]: recovery logic is part of wait.UntilWithContext. it little looks behavior of panic propagation to top is manage be variable ReallyCrash which is set to true by default: https://github.com/kubernetes/apimachinery/blob/master/pkg/util/runtime/runtime.go#L34

File: provider/cmd/app/app.go
Comment (so-sahu) [reply provider/cmd/app/app.go]: Changed

File: pkg/controllers/machine_controller.go
Comment (so-sahu) [reply pkg/controllers/machine_controller.go]: Thanks for your suggestion. It is now followed.

File: provider/cmd/app/app.go
Comment (so-sahu) [reply provider/cmd/app/app.go:123]: I would prefer to mention the link in the description. Something like: `Duration to wait for the VM to gracefully shut down. If the VM does not shut down within this period, it will be forcibly destroyed by garbage collector`

File: pkg/controllers/machine_controller.go
Comment (lukas016) [reply pkg/controllers/machine_controller.go:194]: minimally panic has to be logged in same format as other log records.

garbage collector is for me core functionality of libvirt-provider which has to work. But i prefer here really failure tolerant logic. Main reason is affect another more important logic (create machine). When machine cannot be deleted okay it is problem but you can still create new machines and you have time solve root cause with GC. If panic crash application, app can be stuck in infinity restart loop and i think it is worse case.

Comment (so-sahu) [reply pkg/controllers/machine_controller.go:194]: As discussed privately, there is no need for us to be concerned about this at the moment.

File: provider/cmd/app/app.go
Comment (so-sahu) [reply provider/cmd/app/app.go]: Added it to the description, as well as the name is changed.

File: pkg/controllers/machine_controller.go
Comment (so-sahu) [reply pkg/controllers/machine_controller.go]: removed.

File: pkg/controllers/machine_controller.go
Comment (so-sahu) [reply pkg/controllers/machine_controller.go]: done.

File: pkg/controllers/machine_controller.go
Comment (so-sahu) [reply pkg/controllers/machine_controller.go]: done

File: pkg/controllers/machine_controller.go
Comment (so-sahu) [reply pkg/controllers/machine_controller.go]: changed

File: pkg/controllers/machine_controller.go
Comment (so-sahu) [reply pkg/controllers/machine_controller.go:277]: followed.

Comment (lukas016) [reply pkg/controllers/machine_controller.go:277]: maybe add starting log before every deletion step (volume, nic, machinedir) can be useful for ops guys too

File: provider/cmd/app/app.go
Comment (lukas016) [inline provider/cmd/app/app.go:124]: I suggest here replace duration with interval. Duration could be explained as timeout for GC process.

File: provider/cmd/app/app.go
Comment (lukas016) [inline provider/cmd/app/app.go]: optional: renaming of variables to same design as flags can be useful for programmers.

File: pkg/controllers/machine_controller.go
Comment (so-sahu) [reply pkg/controllers/machine_controller.go:277]: I sense an excessive amount of logging. Nevertheless, the absence of the completed log signifies an issue. Fortunately, we now possess the log indicating the initiation of the function execution.

Comment (lukas016) [reply pkg/controllers/machine_controller.go:277]: ok we can solve it with docs

File: provider/cmd/app/app.go
Comment (so-sahu) [reply provider/cmd/app/app.go]: replaced.

File: provider/cmd/app/app.go
Comment (so-sahu) [reply provider/cmd/app/app.go]: renamed

**Review by lukas016 — COMMENTED** [review]
lgtm i am waiting for rebase before approval

**Review by lukas016 — DISMISSED** [review]

File: pkg/host/store.go
Comment (lukasfrank) [inline pkg/host/store.go:170]: Is there a reason for deleting this?

File: pkg/api/api.go
Comment (lukasfrank) [inline pkg/api/api.go:15]:
```suggestion
	ShutdownAt time.Time  `json:"ShutdownAt"`
```

File: pkg/controllers/machine_controller.go
Comment (lukasfrank) [inline pkg/controllers/machine_controller.go:264]:
```suggestion
			logger := log.WithValues("machineID", machine.ID)
```

**Review by lukasfrank — CHANGES_REQUESTED** [review]

File: pkg/host/store.go
Comment (so-sahu) [reply pkg/host/store.go:170]: Since the garbage collector now fully manages the deletion process, there's no need to enqueue reconciliation. Setting the DeleteAt field alone is now adequate for gracefully shutting down the machine.

File: pkg/api/api.go
Comment (so-sahu) [reply pkg/api/api.go:15]: Presently, we are initiating machine shutdowns at regular intervals. However, in this field, we are preserving the timestamp when the initial shutdown attempt was made. Consequently, I considered that `FirstShutdownAt` might be a more fitting name for this field. Please share your thoughts on this.

File: pkg/controllers/machine_controller.go
Comment (so-sahu) [reply pkg/controllers/machine_controller.go:264]: Ok I will do the change.

File: pkg/controllers/machine_controller.go
Comment (lukas016) [inline pkg/controllers/machine_controller.go:328]:
```suggestion
		if err != nil && !libvirt.IsNotFound(err) {
```

**Review by lukas016 — CHANGES_REQUESTED** [review]

File: pkg/controllers/machine_controller.go
Comment (so-sahu) [reply pkg/controllers/machine_controller.go]: Done

File: pkg/controllers/machine_controller.go
Comment (lukas016) [inline pkg/controllers/machine_controller.go:379]: pls implement graceful destroy

Comment (so-sahu) [reply pkg/controllers/machine_controller.go]: done

File: pkg/controllers/machine_controller.go
Comment (so-sahu) [reply pkg/controllers/machine_controller.go]: changed it to `Destroyed domain`

**Review by lukas016 — DISMISSED** [review]

**Review by lukasfrank — DISMISSED** [review]

File: pkg/controllers/machine_controller.go
Comment (afritzler) [inline pkg/controllers/machine_controller.go:370]: Won't we run the shutdown multiple times here until the shutdown time is reached? Is that the expected behavior?

File: pkg/controllers/machine_controller.go
Comment (afritzler) [inline pkg/controllers/machine_controller.go:324]: Can we add a log here stating the the deletion is triggered?

File: pkg/controllers/machine_controller.go
Comment (afritzler) [inline pkg/controllers/machine_controller.go:400]: Add a log here that the `Shutdown` has been triggered.

**Review by afritzler — CHANGES_REQUESTED** [review]
We might also want to add a test case ensuring the correct behavior e.g. start the service with a relatively short grace period.

File: pkg/controllers/machine_controller.go
Comment (lukas016) [reply pkg/controllers/machine_controller.go:370]: yes it is expected behavior because sometimes signal can be ignore and shutdown call is async call.

Comment (afritzler) [reply pkg/controllers/machine_controller.go:370]: How can a signal be "ignored"? Async should be not a problem here. I was just wondering why we should pounding a shutdown signal to the domain every X seconds.

Comment (lukas016) [reply pkg/controllers/machine_controller.go:370]: pls ask @gehoern, he recommended call it more times (min 2times).

Comment (lukas016) [reply pkg/controllers/machine_controller.go:370]: Async isn't problem and it help us here but it could be problem, if it is blocking call as Destroy.

File: pkg/controllers/machine_controller.go
Comment (lukas016) [reply pkg/controllers/machine_controller.go:400]: you have log before triggering. if it is end with error you will get error message. Isn't this better way how logs stuff? you know what was current operation and you don't need additional fields for errors.

Comment (so-sahu) [reply pkg/controllers/machine_controller.go:400]: Is it necessary in this context? Given that we are already logging `Triggering shutdown`, the absence of any subsequent error logs implies it's triggered.

File: pkg/controllers/machine_controller.go
Comment (so-sahu) [reply pkg/controllers/machine_controller.go:324]: We have the option to include it. However, this log will be reiterated in the console following every resync interval, in conjunction with the "Triggering shutdown" log. We already possess logs that clearly illustrate whether a shutdown or domain destruction has taken place. Is this the anticipated behavior?

**Comment by so-sahu** [comment]
> We might also want to add a test case ensuring the correct behavior e.g. start the service with a relatively short grace period.

While the integration tests are currently in progress, it would be nice to include these tests there to prevent any potential conflicts.

File: pkg/controllers/machine_controller.go
Comment (lukasfrank) [reply pkg/controllers/machine_controller.go:370]: The `VM` can miss the `AcpiPowerBtn` signal in case of heavy load. Therefore the signal is sent multiple times. Maybe put a comment in the code to document it.

Comment (so-sahu) [reply pkg/controllers/machine_controller.go:370]: Comment added.

Comment (afritzler) [reply pkg/controllers/machine_controller.go:370]: Fair enought. Then I would suggest to put some comments in the code explaining this circumstances.

Comment (afritzler) [reply pkg/controllers/machine_controller.go:370]: I see this has been addressed already.

**Review by lukas016 — APPROVED** [review]

**Review by lukasfrank — APPROVED** [review]

**Review by afritzler — APPROVED** [review]

**Pull request merged**
